File: baekjoon/1717.py
from sys import stdin

def find(a):
    if parent[a] == a:
        return a
    else:
        parent[a] = find(parent[a])
        return parent[a]

def union(a, b):
    a = find(a)
    b = find(b)
    if a != b:
        parent[b] = a

n, m = map(int, stdin.readline().split())
parent = [i for i in range(n+1)]

for _ in range(m):
    x, a, b = map(int, stdin.readline().split())
    if x == 0:
        union(a, b)
    elif x == 1:
        if find(a) == find(b):
            print("YES")
        else:
            print("NO")

File: baekjoon/1717.cpp
#include <iostream>
#include <vector>
using namespace std;

vector<int> parent;

int find(int a) {
    if (parent[a] == a) return a;
    else return parent[a] = find(parent[a]);
}

bool unite(int a, int b) {
    a = find(a);
    b = find(b);
    if (a == b) return false;
    parent[b] = a;
    return true;
}

void solve() {
    int n, m;
    cin >> n >> m;

    parent = vector<int>(n + 1);
    for (int i = 0; i < n + 1; i++) parent[i] = i;

    while (m--) {
        int q, a, b;
        cin >> q >> a >> b;

        if (q == 0) {
            unite(a, b);
        } else if (q == 1) {
            if (find(a) == find(b)) cout << "YES\n";
            else cout << "NO\n";
        }
    }
}

int main() {
    ios_base::sync_with_stdio(0);
    cin.tie(0);
    cout.tie(0);

    int T = 1;
    for (int i = 0; i < T; i++) {
        solve();
    }
}